### DISCUSSION TITLE
Is there a way to acquire driver gaps at the end of each lap?

### DISCUSSION THREAD

**Discussion by laurence9899** [theOehrly/Fast-F1#503] [Q&A] [answered]
Hi I'm working on acquiring a dataset for a Machine learning strategy model for a University project and I am trying to get the lap by lap intervals between each drivers. I have used fastf1.api.timing_data and  the stream_data object but it gives that data for multiple samples per lap. I was just wondering if there's a good way of extracting this information for the end of each lap?

Thanks 
Laurence

**Comment by theOehrly** [comment, marked as answer]
One of those samples is always taken at the finish line. They timestamps line up reasonably well with the timing data. Therefore you can use the sample with the smallest absolute difference between its timestamp and the time at which the lap was set.

I'm using the following code internally for the same problem:

```
def _get_gap_str_for_drv(drv, idx, laps_data, stream_data):
    first_time = laps_data[laps_data['Driver'] == drv].iloc[idx]['Time']
    ref_idx = (stream_data[stream_data['Driver'] == drv]['Time']
               - first_time).abs().idxmin()
    gap_str = stream_data.loc[ref_idx]['GapToLeader']
    return gap_str
```

**Reply by laurence9899** [reply]
Thank you very much, that works for what I need.
